### PR TITLE
Merge ffi-io branch into master

### PR DESCRIPTION
**Pull request by chuckremes**
Here's a pull request to start discussing merging this branch into master. Let's talk code style, refactoring, specs, or whatever else is relevant.

### CODE REVIEW

File: kernel/delta/io.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Redefine STDIN, STDOUT & STDERR to use the new IO class. It reopened and redefined
+# all methods used in the bootstrap step. Secondly, update the $std* globals to point
+# to the new objects.
+
+def redefine_io(fd, mode)
```
Comment (YorickPeterse) [inline kernel/delta/io.rb]: This method should eventually most likely be in a mirror, or scoped somewhere under the `Rubinius` namespace in general. Right now it's defined at the top-level, which is likely to clash with other methods.

File: spec/default.mspec [hunk 2]
```diff
@@ -46,4 +46,13 @@ class MSpecScript
       MSpec.enable_feature :generator
     end
   end
+  
+  if IO.const_defined?(:POSIX_FADVISE_NORMAL) && IO.const_defined?(:POSIX_FADVISE_DONTNEED)
+  	# Next, check to make sure both those values aren't zero; if they are, then this OS doesn't
+	# support it. This is a hack and should probably be improved as platform and feature 
+	# detection changes in the build process
+	unless IO::POSIX_FADVISE_NORMAL.zero? && IO::POSIX_FADVISE_DONTNEED.zero?
+		MSpec.enable_feature :posix_fadvise
```
Comment (YorickPeterse) [inline spec/default.mspec]: Indentation is wrong

File: spec/ruby/core/io/write_spec.rb [hunk 3]
```diff
@@ -22,7 +22,7 @@
 
   # TODO: impl detail? discuss this with matz. This spec is useless. - rdavis
   # I agree. I've marked it not compliant on macruby, as we don't buffer input. -pthomson
```
Comment (YorickPeterse) [inline spec/ruby/core/io/write_spec.rb]: While you're working on this file, nuke these comments, they're pointless to keep around.

File: vm/builtin/io.cpp [hunk 4]
```diff
@@ -1517,4 +1518,54 @@ namespace rubinius {
...
+    }
+    else {
+      return cFalse;
+    }
+  }
+
+  Object* FDSet::to_set(STATE) {
+	void *ptr = (void*)&descriptor_set;
+
+	return Pointer::create(state, ptr);
```
Comment (YorickPeterse) [inline vm/builtin/io.cpp]: Indentation appears to be broken here.

**Comment by YorickPeterse** [comment]
One general request in advance: please don't use single letter variables. Looking at code I have no clue what `n` means. For small methods it's not a huge issue, but I've seen plenty of methods grow to certain sizes where it's an absolute nightmare trying to figure out what `i`, `n`, `a` and what not all represent.

**Comment by chuckremes** [comment]
Fixed a bunch of these.

**Comment by chuckremes** [comment]
The latest commits for the C-API have some whitespace issues. I'll get them fixed. In the meantime, need help debugging it.

**Comment by chuckremes** [comment]
@brixen did a bunch of work to merge the codedb branch into the ffi-io branch, so we'll use that going forward. This PR will be closed and I'll open one for merging the codedb-ffi-io branch into master. That long branch name will encourage us to merge it quickly to save precious keystrokes!

**Pull request closed without merging**